clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp

```diff
@@ -252,6 +259,51 @@ Error IncludeTreeActionController::finalize(CompilerInstance &ScanInstance,
   return Error::success();
 }
 
+Error IncludeTreeActionController::initializeModuleBuild(
+    CompilerInstance &ModuleScanInstance) {
+  BuilderStack.push_back(
+      std::make_unique<IncludeTreeBuilder>(DB, PrefixMapper));
+
+  // Attach callbacks for the IncludeTree of the module. The preprocessor
+  // does not exist yet, so we need to indirect this via DependencyCollector.
+  auto DC = std::make_shared<PPCallbacksDependencyCollector>(
+      [&Builder = current()](Preprocessor &PP) {
+        return std::make_unique<IncludeTreePPCallbacks>(Builder, PP);
+      });
+  ModuleScanInstance.addDependencyCollector(std::move(DC));
+
+  return Error::success();
+}
+
+Error IncludeTreeActionController::finalizeModuleBuild(
+    CompilerInstance &ModuleScanInstance) {
+  // FIXME: the scan invocation is incorrect here; we need the `NewInvocation`
+  // from `finalizeModuleInvocation` to finish the tree.
+  auto Builder = BuilderStack.pop_back_val();
+  auto Tree = Builder->finishIncludeTree(ModuleScanInstance,
+                                         ModuleScanInstance.getInvocation());
+  if (!Tree)
+    return Tree.takeError();
+
+  Module *M = ModuleScanInstance.getPreprocessor().getCurrentModule();
+  assert(M && "finalizing without a module");
+  M->setIncludeTreeID(Tree->getID().toString());
+
+  return Error::success();
+}
+
+Error IncludeTreeActionController::finalizeModuleInvocation(
+    CompilerInvocation &CI, const ModuleDeps &MD) {
+  if (auto ID = MD.IncludeTreeID) {
```

Comment:
Is it valid for `MD.IncludeTreeID` to be `nullopt`? If yes could you add a comment clarifying when that is?

Reply:
Good point. I made it return an error; it probably "shouldn't happen", but it's not something easy to prove that it can't happen, so I didn't want to only have an assertion.